[1.12.x] Bumped the biome limit to 65,536

Madsthunder opened the pull request:
This pr _should_ maintain binary compatibility, so long as there aren't more than 256 biomes registered **_and_** there isn't a mod present that tinkers with biomes. Resolves #1704.

Madsthunder committed:
Pushed the biome limit from 256 to 65,536

covers1624 commented on patches/minecraft/net/minecraft/network/play/server/SPacketChunkData.java.patch:
```diff
@@ -0,0 +1,18 @@
+         if (this.func_149274_i())
+         {
+             p_189555_1_.writeBytes(p_189555_2_.func_76605_m());
++            int localBiomeCount = p_189555_2_.getLocalBiomes().size(); p_189555_1_.writeByte((byte)(localBiomeCount & 255)); Short[] localBiomeIds = p_189555_2_.getLocalBiomeIds(); for (int k = 0; k < localBiomeCount; k++) p_189555_1_.writeShort(localBiomeIds[i]);
```
lolwot, you don't need to squish it to one line, make it readable.

jabelar replied:
Actually the official guide for contributing says the opposite: https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge 

"In spirit of the previous point: prevent to add more lines than necessary. If you can reduce your statement to one single line, please do so. "

They actually use an example where they basically say that it will reduce readability.  And even though they say "statement" their example is a nested if statement that they squash.

HenryLoenwind replied:
I never really understood that guideline. For merging it shouldn't matter how many lines you add---in one chunk. Merging a 1000-line addition into the right location is just the same as merging a 1-line addition to that location and about 1/1000th the work of merging 1000 1-line additions.

covers1624 commented on patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch:
```diff
@@ -258,7 +258,15 @@
+         if (p_186033_3_)
+         {
+             p_186033_1_.readBytes(this.field_76651_r);
++            this.localBiomeIds = new Short[256]; this.localBiomes.clear(); if (p_186033_1_.readableBytes() > 0) { int localBiomeCount = p_186033_1_.readByte() + 128; for (int j = 0; j < localBiomeCount; j++) { short localBiomeId = p_186033_1_.readShort(); this.localBiomeIds[this.localBiomes.size()] = localBiomeId; this.localBiomes.add(Biome.func_150568_d(localBiomeId & 65535)); } }
```
and here

covers1624 commented on patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch:
```diff
@@ -275,7 +283,20 @@
+         {
+             Biome biome = p_177411_2_.func_180300_a(p_177411_1_, Biomes.field_76772_c);
+             k = Biome.func_185362_a(biome);
++            int localBiomeIndex = this.localBiomes.indexOf(biome); if (localBiomeIndex < 0) { localBiomeIndex = this.localBiomes.size(); this.localBiomes.add(biome); this.localBiomeIds[localBiomeIndex] = (short)(localBiomeIndex); } k = localBiomeIndex;
```
and here

covers1624 commented on patches/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java.patch:
```diff
@@ -97,7 +97,15 @@
+ 
+         p_75820_3_.func_74782_a("Sections", nbttaglist);
+         p_75820_3_.func_74773_a("Biomes", p_75820_1_.func_76605_m());
++        int localBiomeCount = p_75820_1_.getLocalBiomes().size(); Short[] localBiomeIds = p_75820_1_.getLocalBiomeIds(); byte[] localBiomeArray = new byte[localBiomeCount * 2]; for (int i = 0; i < localBiomeCount; i++) { localBiomeArray[i] = (byte)(localBiomeIds[i] / 256); localBiomeArray[i] = (byte)(localBiomeIds[i] & 255); } p_75820_3_.func_74773_a("LocalBiomes", localBiomeArray);
```
and here

covers1624 commented on patches/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java.patch:
```diff
@@ -153,10 +161,11 @@
              chunk.func_76616_a(p_75823_2_.func_74770_j("Biomes"));
          }
  
++        if (p_75823_2_.func_150297_b("LocalBiomes", 7)) { List<net.minecraft.world.biome.Biome> localBiomes = chunk.getLocalBiomes(); localBiomes.clear(); Short[] localBiomeIds = chunk.getLocalBiomeIds(); byte[] localBiomeArray = p_75823_2_.func_74770_j("LocalBiomes"); for (int i1 = 0; i1 < localBiomeArray.length / 2; i1++) { localBiomes.add(net.minecraft.world.biome.Biome.func_180276_a((localBi ...
```
and here..

Madsthunder committed:
Formatting [1]

Replaced stray tabs with quadruple-spaces
Spread the one-line patches made in SPackedChunkData, Chunk, and AnvilChunkLoader across multiple lines

williewillus commented:
And how does a vanilla client handle this?

covers1624 commented:
I intended to put comments on all your patches, but there are soo many. (still more formatting to change.)

Madsthunder committed:
Fixes [1]

Fixed NPE being thrown when sending chunk packets to client
Fixed biome saving
Fixed biomes from newly generated worlds being either an ocean, plains, or desert

Madsthunder commented:
@williewillus Before the last commit, the vanilla client would handle this just fine on a **_pre-generated world_**. With newly generated worlds, the server would send the _localized_ version of the biome array rather than the _vanilla_ version.

liach commented:
this one bresks all pre existing world generators, right?

Madsthunder commented:
Why was this closed

@liach Yes, it does if there are more than 255 biomes registered

Dockter commented:
@Liach where as it would break TC to pieces, it would be an easy thing to fix.  This would solve a long running issue with TC and how it implements biomes at the client level.

liach commented:
I think this may be too big for 1.12. I'd leave it to 1.13.

Dockter commented:
@liach I predict mods won't see a large move to 1.13 until the end of 2018 simply due to the enormous amount of time its going to take to adapt to the 1.13 ecosystem.

Madsthunder commented:
Understandable. I'll try to remake this for 1.13 after I remake #3664.

williewillus commented:
My question above was, if a forge server with this patch accepts a connection from a vanilla client, will the vanilla client crash horribly (or leak memory everywhere)?

I'm not sure how it responds to packet sizes being mismatched, but if since you modify the length of the chunk data packet that can break a lot of things

Madsthunder commented:
@williewillus That's what I meant above, I think I should've clarified. The client just accepts the chunk data packet from the server as an oversized packet. The client will stop at extracting biomes and carry on processing other chunk packets like normal.
